### PR TITLE
Refactor transform_messages

### PR DESCRIPTION
**Pull request by dkirsche**
## Why are these changes needed?
Some small refactors to improve readabilty, simplify and optimize.

Improvements
* The refactored code reduces the overall complexity and the number of iterations over the messages. 
* Consolidated Truncation Logic: The logic for truncating each message to max_tokens_per_message and trimming the list to max_messages is more straightforward and easier to follow. It consolidates several steps into fewer lines of code.
* Total tokens count excludes system to be apples to apples comparison to processed_messages_tokens
* truncate_str_to_tokens requires less iterations. Iterate over tokens instead of characters.

## Related issue number
None

## Checks

- [X ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by dkirsche** [comment]
> @dkirsche please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

**Comment by ekzhu** [comment]
@dkirsche  thanks for the PR!! It looks like you are digging deep into this project to make this change request. Could you share a little bit about your usage scenario? You are welcome to join our discord we can add you to dev role once the PR is merged.

**Comment by gagb** [comment]
@dkirsche nice PR! Left some comments;

I like the heuristic to iterate over words instead of characters; Tho do you think that would upset the formatting of the OG message? Maybe we can add a test for that?

File: autogen/agentchat/contrib/capabilities/context_handling.py
Comment (afourney) [inline autogen/agentchat/contrib/capabilities/context_handling.py:117]: Won't this break right away?  actual_token_count will start out smaller than max_tokens

Comment (dkirsche) [reply autogen/agentchat/contrib/capabilities/context_handling.py:117]: This function is calculating tokens by splitting words, which is different than how token_count_utils.count_token() works. This means that the two methods might have slightly different values. This if statement is an extra guard to make sure that the simplistic calculation will never be larger than token_count_utils.count_token. If its larger, than the loop will decrement by string by one token and do the evaluation again. 
If its slightly smaller, I'm ok with that since its below the max threshold.  
Working on some tests for this.

File: autogen/agentchat/contrib/capabilities/context_handling.py
Comment (jackgerrits) [inline autogen/agentchat/contrib/capabilities/context_handling.py:116]: Since we can use tiktoken to actually tokenize the input wouldnt it make sense to use that to truncate rather than tokenizing it over and over and calculating the length?

Comment (afourney) [reply autogen/agentchat/contrib/capabilities/context_handling.py:116]: Can we tokenize it, truncate the token sequence, then transform it back to text? If so, that might work... but we'll have to be a little careful about word boundaries (since a token can very-often be sub-word)

Comment (jackgerrits) [reply autogen/agentchat/contrib/capabilities/context_handling.py:116]: Oof yeah you're right. That's a real pain

Comment (dkirsche) [reply autogen/agentchat/contrib/capabilities/context_handling.py:116]: I'm wondering if we even need to worry about tokenizing. Models can tokenize slightly differently, so it can vary depending on the model and token_count_utils.count_token is defaulting to gpt3.5. I could simplify this tremendously if we are good with approximating tokens via `text.split()`. The five lines in the for loop becomes one line `truncated_tokens = tokens[:max_tokens]`. I would prefer to simplify it.

Comment (jackgerrits) [reply autogen/agentchat/contrib/capabilities/context_handling.py:116]: There is often more than 1 token per word, so if we went with that scheme it would differ a lot. I think that matters especially for handling what we do at that limit. You're right the current method is definitely suboptimal. For now I think what you have is good but I can see us adding functionality to tokenize based on the model being used

Comment (dkirsche) [reply autogen/agentchat/contrib/capabilities/context_handling.py:116]: That makes sense. It would be a valuable improvement.

**Comment by dkirsche** [comment]
If all looks ok, can I get approval to merge?

**Comment by dkirsche** [comment]
@ekzhu @gagb Let me know if there is anything else needed for this PR. Lets get it shipped! Thanks

File: test/agentchat/contrib/capabilities/test_context_handling.py
Comment (gagb) [reply test/agentchat/contrib/capabilities/test_context_handling.py:191]: @dkirsche did you look at this comment?

Comment (dkirsche) [reply test/agentchat/contrib/capabilities/test_context_handling.py:191]: Good point. I have added that test case and update the function to more directly calculate tokens by using tiktoken to encode and then decode which was a great idea that was suggested earlier.

**Comment by gagb** [comment]
@dkirsche please look at the commits I submitted here: https://github.com/dkirsche/autogen/pull/1

Especially pay attention to the second commit which added a new test case and fixed a bug. Lmk what you think!

**Comment by dkirsche** [comment]
> @dkirsche please look at the commits I submitted here: [dkirsche#1](https://github.com/dkirsche/autogen/pull/1)
> 
> Especially pay attention to the second commit which added a new test case and fixed a bug. Lmk what you think!

Your changes made sense. I pulled them into the PR.

**Review by gagb — APPROVED** [review]

**Comment by gagb** [comment]
> > @dkirsche please look at the commits I submitted here: [dkirsche#1](https://github.com/dkirsche/autogen/pull/1)
> > Especially pay attention to the second commit which added a new test case and fixed a bug. Lmk what you think!
> 
> Your changes made sense. I pulled them into the PR.

Thank you so much @dkirsche! Also, I think we need to improve this function to handle the case when the message history has tool_calls and responses. But we can fix this in a new PR.

**Comment by dkirsche** [comment]
Some of the checks are in a waiting state. @ekzhu I believe I need your review before proceeding.

**Pull request merged**
